src/charm.py

```diff
@@ -137,76 +155,105 @@ def _update_layer(self) -> None:
+        )
+
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
             event.defer()
-            logger.info("Missing database details. Deferring the event.")
-            self.unit.status = WaitingStatus("Waiting for database creation")
+            self.unit.status = WaitingStatus("Waiting for Hydra service")
+            logger.info("Hydra service is absent. Deferring database created event.")
+            return
+
+        logger.info("Updating Hydra config and restarting service")
+        self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+
+        try:
+            self._run_sql_migration(set_timeout=True)
+        except ExecError as err:
+            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
+            self.unit.status = BlockedStatus("Database migration job failed")
+            logger.error("Automigration job failed, please use the run-migration action")
             return
 
+        self._container.start(self._container_name)
+        self.unit.status = ActiveStatus()
+
+    def _on_database_changed(self, event) -> None:
+        """Event Handler for database changed event."""
         if not self._container.can_connect():
             event.defer()
             logger.info("Cannot connect to Hydra container. Deferring the event.")
             self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
             return
 
-        self._update_layer()
+        self.unit.status = MaintenanceStatus("Updating database details")
 
-        if not self.unit.is_leader():
-            # TODO: Observe leader_elected event
-            logger.info("Unit does not have leadership")
-            self.unit.status = WaitingStatus("Waiting for leadership")
+        try:
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            event.defer()
+            self.unit.status = WaitingStatus("Waiting for Hydra service")
+            logger.info("Hydra service is absent. Deferring database created event.")
             return
 
-        self._run_sql_migration()
-
+        logger.info("Updating Hydra config and restarting service")
+        self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+        self._container.restart(self._container_name)
         self.unit.status = ActiveStatus()
 
-    def _run_sql_migration(self) -> None:
-        """Runs a command to create SQL schemas and apply migration plans."""
-        process = self._container.exec(
-            ["hydra", "migrate", "sql", "-e", "--config", self._hydra_config_path, "--yes"],
-            timeout=20.0,
-        )
+    def _on_run_migration(self, event: ActionEvent) -> None:
+        """Runs the migration as an action response."""
+        if not self.unit.is_leader():
+            event.fail("The action can be run only on leader unit")
+            return
```

Comment:
If the user wants to run the migration from a non-leader unit, I see no reason to stop them.